File: scripts/workspace-launcher/ghostty-window-grid.py
#!/usr/bin/env python3
"""
Ghostty Window Grid Layout
Uses Quartz/AppKit to position windows directly
"""

import Quartz
import AppKit
from Foundation import NSMakeRect
import sys
import time

def get_ghostty_windows():
    """Get all Ghostty windows using Quartz window server"""
    window_list = Quartz.CGWindowListCopyWindowInfo(
        Quartz.kCGWindowListOptionOnScreenOnly | Quartz.kCGWindowListExcludeDesktopElements,
        Quartz.kCGNullWindowID
    )

    ghostty_windows = []
    for window in window_list:
        owner_name = window.get(Quartz.kCGWindowOwnerName, '')
        window_layer = window.get(Quartz.kCGWindowLayer, -1)
        window_name = window.get(Quartz.kCGWindowName, 'Untitled')

        # Find Ghostty windows (standard windows on layer 0)
        if owner_name.lower() == 'ghostty' and window_layer == 0:
            window_number = window.get(Quartz.kCGWindowNumber)
            bounds = window.get(Quartz.kCGWindowBounds)
            ghostty_windows.append({
                'number': window_number,
                'name': window_name,
                'bounds': bounds,
                'pid': window.get(Quartz.kCGWindowOwnerPID)
            })

    return ghostty_windows

def get_screen_dimensions():
    """Get main screen dimensions"""
    screen = AppKit.NSScreen.mainScreen()
    frame = screen.frame()
    return frame.size.width, frame.size.height

def position_windows_grid(windows, cols=3, rows=2, gap=10, top_margin=60):
    """Position windows in a grid layout"""
    screen_width, screen_height = get_screen_dimensions()

    print(f"Screen dimensions: {screen_width:.0f}x{screen_height:.0f}")
    print(f"Found {len(windows)} Ghostty windows")

    # Calculate usable area
    usable_width = screen_width
    usable_height = screen_height - top_margin

    # Calculate window dimensions
    total_gap_width = gap * (cols - 1)
    total_gap_height = gap * (rows - 1)
    window_width = (usable_width - total_gap_width) / cols
    window_height = (usable_height - total_gap_height) / rows

    print(f"Window size: {window_width:.0f}x{window_height:.0f}")

    # Grid positions
    grid_positions = [
        (0, 0),  # CHRONOS
        (0, 1),  # IMAGINARIUM
        (0, 2),  # ARCHITECTUS
        (1, 0),  # LUDUS
        (1, 1),  # OCULUS
        (1, 2),  # OPERATUS
    ]

    # Use accessibility API to move windows
    app = AppKit.NSRunningApplication.runningApplicationsWithBundleIdentifier_(
        'com.mitchellh.ghostty'
    )

    if not app:
        print("❌ Could not find Ghostty application")
        return False

    app = app[0]
    app.activateWithOptions_(AppKit.NSApplicationActivateIgnoringOtherApps)
    time.sleep(0.5)

    # Try to position windows using AppleScript as fallback
    for i, (window, (row, col)) in enumerate(zip(windows[:6], grid_positions)):
        window_x = col * (window_width + gap)
        window_y = top_margin + row * (window_height + gap)

        print(f"Positioning window {i+1} ({window['name']}) at ({window_x:.0f}, {window_y:.0f})")

        # Use osascript to move the window
        script = f'''
tell application "System Events"
    tell process "ghostty"
        try
            set position of window {i+1} to {{{int(window_x)}, {int(window_y)}}}
            set size of window {i+1} to {{{int(window_width)}, {int(window_height)}}}
        on error errMsg
            log "Error moving window {i+1}: " & errMsg
        end try
    end tell
end tell
'''
        import subprocess
        try:
            subprocess.run(['osascript', '-e', script], capture_output=True, timeout=2)
        except Exception as e:
            print(f"  Warning: Could not position window {i+1}: {e}")

    return True

def main():
    windows = get_ghostty_windows()

    if not windows:
        print("No Ghostty windows found")
        return 1

    success = position_windows_grid(windows)

    if success:
        print("\n✅ Window grid layout complete!")
        print("\nGrid layout:")
        print("┌─────────────┬─────────────┬─────────────┐")
        print("│   CHRONOS   │ IMAGINARIUM │ ARCHITECTUS │")
        print("├─────────────┼─────────────┼─────────────┤")
        print("│    LUDUS    │   OCULUS    │  OPERATUS   │")
        print("└─────────────┴─────────────┴─────────────┘")
        return 0
    else:
        return 1

if __name__ == '__main__':
    sys.exit(main())

File: scripts/workspace-launcher/ghostty-window-grid.js
#!/usr/bin/osascript -l JavaScript
/**
 * Ghostty Window Grid Layout (JXA version)
 * Uses JavaScript for Automation to position windows
 */

ObjC.import('Cocoa');
ObjC.import('stdlib');

function _run(_argv) {
  // Configuration
  const gridCols = 3;
  const gridRows = 2;
  const windowGap = 10;
  const topMargin = 60;
  const bottomMargin = 0;
  const leftMargin = 0;
  const rightMargin = 0;

  // Get screen dimensions
  const screen = $.NSScreen.mainScreen;
  const screenFrame = screen.frame;
  const screenWidth = screenFrame.size.width;
  const screenHeight = screenFrame.size.height;

  console.log(`Screen dimensions: ${screenWidth}x${screenHeight}`);

  // Calculate usable area
  const usableX = leftMargin;
  const usableY = topMargin;
  const usableWidth = screenWidth - leftMargin - rightMargin;
  const usableHeight = screenHeight - topMargin - bottomMargin;

  // Calculate window dimensions
  const totalGapWidth = windowGap * (gridCols - 1);
  const totalGapHeight = windowGap * (gridRows - 1);
  const windowWidth = (usableWidth - totalGapWidth) / gridCols;
  const windowHeight = (usableHeight - totalGapHeight) / gridRows;

  console.log(`Window size: ${windowWidth}x${windowHeight}`);

  // Get all windows from window server
  const windowList = $.CGWindowListCopyWindowInfo(
    $.kCGWindowListOptionOnScreenOnly | $.kCGWindowListExcludeDesktopElements,
    $.kCGNullWindowID,
  );

  const windowCount = $.CFArrayGetCount(windowList);
  console.log(`Total windows on screen: ${windowCount}`);

  // Find Ghostty windows
  const ghosttyWindows = [];

  for (let i = 0; i < windowCount; i++) {
    const windowInfo = $.CFArrayGetValueAtIndex(windowList, i);
    const ownerName = ObjC.unwrap($.CFDictionaryGetValue(windowInfo, $.kCGWindowOwnerName));
    const windowName = ObjC.unwrap($.CFDictionaryGetValue(windowInfo, $.kCGWindowName));
    const windowLayer = ObjC.unwrap($.CFDictionaryGetValue(windowInfo, $.kCGWindowLayer));

    // Look for Ghostty windows (owner name is "ghostty")
    if (ownerName && ownerName.toLowerCase() === 'ghostty' && windowLayer === 0) {
      const windowNumber = ObjC.unwrap($.CFDictionaryGetValue(windowInfo, $.kCGWindowNumber));
      ghosttyWindows.push({
        number: windowNumber,
        name: windowName || 'Untitled',
        info: windowInfo,
      });
    }
  }

  console.log(`Found ${ghosttyWindows.length} Ghostty windows`);

  if (ghosttyWindows.length === 0) {
    console.log('No Ghostty windows found');
    return;
  }

  // Grid positions for each window
  const gridPositions = [
    [0, 0], // CHRONOS
    [0, 1], // IMAGINARIUM
    [0, 2], // ARCHITECTUS
    [1, 0], // LUDUS
    [1, 1], // OCULUS
    [1, 2], // OPERATUS
  ];

  // Position each window using AXUIElement
  const _systemWide = $.AXUIElementCreateSystemWide();

  for (let i = 0; i < Math.min(ghosttyWindows.length, 6); i++) {
    const win = ghosttyWindows[i];
    const gridPos = gridPositions[i];
    const gridRow = gridPos[0];
    const gridCol = gridPos[1];

    // Calculate window position
    const windowX = usableX + gridCol * (windowWidth + windowGap);
    const windowY = usableY + gridRow * (windowHeight + windowGap);

    console.log(`Positioning window ${i + 1} (${win.name}) at (${Math.round(windowX)}, ${Math.round(windowY)})`);

    // Try to position the window using AXUIElement
    const _windowElement = $.AXUIElementCreateApplication($.getpid());

    // Note: This approach requires finding the window by its window number
    // and using accessibility API, which Ghostty might not fully support

    // Alternative: Use AppleScript bridge
    try {
      const app = Application.currentApplication();
      app.includeStandardAdditions = true;

      // Use osascript to move window via System Events
      app.doShellScript(`osascript -e 'tell application "System Events"
        tell process "ghostty"
          try
            set position of window ${i + 1} to {${Math.round(windowX)}, ${Math.round(windowY)}}
            set size of window ${i + 1} to {${Math.round(windowWidth)}, ${Math.round(windowHeight)}}
          end try
        end tell
      end tell'`);
    } catch (e) {
      console.log(`Could not position window ${i + 1}: ${e}`);
    }
  }

  console.log('✅ Window grid layout complete!');
}
